auraed/src/runtime/cell_service/executables/executable.rs

```diff
@@ -23,66 +38,179 @@ enum ExecutableState {
 
         Ok(())
     }
 
     /// Stops the executable and returns the [ExitStatus].
     /// If the executable has never been started, returns [None].
-    pub fn kill(&mut self) -> Result<Option<ExitStatus>, io::Error> {
-        match &mut self.state {
-            ExecutableState::Init { .. } => Ok(None),
-            ExecutableState::Started { process, .. } => {
-                process.kill(Some(SIGKILL))?;
-                let exit_status = process.wait()?;
-                self.state = ExecutableState::Stopped(exit_status);
-                Ok(Some(exit_status))
-            }
-            ExecutableState::Stopped(exit_status) => Ok(Some(*exit_status)),
+    pub fn kill(&mut self) -> io::Result<Option<ExitStatus>> {
+        let exit_status: Option<ExitStatus>;
+        {
+            let mut inner = self.inner.lock().map_err(|e| {
+                io::Error::new(io::ErrorKind::Other, e.to_string())
+            })?;
+            match &mut inner.state {
+                ExecutableState::Init { .. } => exit_status = None,
+                ExecutableState::Started { process, .. } => {
+                    let proc_status: ExitStatus;
+                    {
+                        let mut proc = process.lock().map_err(|e| {
+                            io::Error::new(io::ErrorKind::Other, e.to_string())
+                        })?;
+                        proc.kill(Some(SIGKILL))?;
+                        proc_status = proc.wait()?;
+                    }
+                    inner.state = ExecutableState::Stopped(proc_status);
+                    exit_status = Some(proc_status);
+                }
+                ExecutableState::Stopped(status) => exit_status = Some(*status),
+            };
         }
+        self.log_thread
+            .take()
+            .map(thread::JoinHandle::join)
+            .expect("thread panicked")
+            .expect("join log_thread")
+            .map_err(|e| io::Error::new(io::ErrorKind::Other, e.to_string()))?;
+        Ok(exit_status)
     }
 
     /// Returns the [Pid] while [Executable] is running, otherwise returns [None].
-    pub fn pid(&self) -> Option<Pid> {
-        let ExecutableState::Started { process, .. } = &self.state else {
-            return None;
+    pub fn pid(&self) -> io::Result<Option<Pid>> {
+        let inner = self
+            .inner
+            .lock()
+            .map_err(|e| io::Error::new(io::ErrorKind::Other, e.to_string()))?;
+        let ExecutableState::Started { process, .. } = &inner.state else {
+            return Ok(None);
         };
+        let proc = process
+            .lock()
+            .map_err(|e| io::Error::new(io::ErrorKind::Other, e.to_string()))?;
 
-        Some(process.pid())
+        Ok(Some(proc.pid()))
     }
 }
 
+/// Spawns a thread that produces log lines while the [Executable] is running.
+fn spawn_log_thread(
+    inner: Arc<Mutex<ExecutableInner>>,
+) -> thread::JoinHandle<io::Result<()>> {
+    let local_inner = inner;
```

Comment:
This doesn't seem needed. `inner` should move into the closure without issue. No?

Reply:
yeah, this was a holdover.